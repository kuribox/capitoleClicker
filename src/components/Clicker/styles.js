import { StyleSheet } from 'react-native';
import variables from '../../theme/variables';

export default StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-between',
    backgroundColor: variables.backgroundColor,
    padding: variables.padding
  },
  clickerButton: {
    backgroundColor: variables.primary,
    borderColor: '#32507c',
    borderWidth: 10,
    borderRadius: 60,
    width: '100%',
    marginVertical: variables.padding,
  },
  clickerValue: {
    textAlign: "center",
    fontSize: 35,
    color: variables.secondary,
    padding: variables.padding,
  },
  buttonCountainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
  },
  legend: {
    fontSize: 18,
  },
  successContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
  },
  successText: {
    fontSize: 40,
    fontWeight: 'bold'
  }
});
