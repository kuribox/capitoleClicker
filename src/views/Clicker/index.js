import React, { useState, useEffect } from 'react';
import { Button, StyleSheet, Text, View } from 'react-native';
import ClickerComponent from '../../components/Clicker';
import { useStore } from '../../providers/StoreContext';

const Clicker = ({navigation})  => {
  const { initUser } = useStore();
  const [user, setUser] = useState(null);

  useEffect(() => {
    initUser().then(userData => {
      if (userData) {
        setUser(userData)

        navigation.setOptions({
          headerTitle: `Clicker - ${userData.name}`,
        });
      }
    })
  }, []);

  if (!user) return null;

  return <ClickerComponent user={user} />;
}

export default Clicker;