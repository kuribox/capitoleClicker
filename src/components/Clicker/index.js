import React, { useLayoutEffect, useState, useEffect } from 'react';
import { Button, Text, View, TouchableOpacity } from 'react-native';
import useInterval from '../../utils/useInterval';
import config from '../../config/config';
import AutoClicker from '../AutoClicker';
import { useStore } from '../../providers/StoreContext';
import styles from "./styles";

const Clicker = ({ user })  => {
  const { updateUser } = useStore();

  const [clickerValue, setClickerValue] = useState(0);
  const [success, setSuccess] = useState(0);

  const [autoClickers, setAutoclickers] = useState([]);

  useEffect(() => {
    if (user) {
      setClickerValue(user.value);
      setAutoclickers(user.clickers);
    }
  }, []);

  useEffect(() => updateUser({ ...user, value: clickerValue, clickers: autoClickers }), [clickerValue, autoClickers]);

  useInterval(() => {
    if (autoClickers.length <= 0) return;
    incrementClicker(autoClickers.reduce((p, c) => p + c.value, 0))
  }, config.clicker.timeout);

  const incrementClicker = (amount = 1) => {
    if (clickerValue + amount > Number.MAX_SAFE_INTEGER) {
      setSuccess(true);
    } else {
      setClickerValue(clickerValue + amount);
    }
  }

  const substractClicker = (value) => {
    if (clickerValue >= value) {
      setClickerValue(clickerValue - value);
      return true;
    }
    return false;
  }

  // create clicker
  const createNewClicker = (name, buyAmount, clickAmount) => {
    if (substractClicker(buyAmount)) {
      setAutoclickers([
        ...autoClickers,
        { id: `${name}_${autoClickers.length}`, name, value: clickAmount }
      ]);
    }
  }

  if (success) return  (
    <View style={styles.successContainer}>
      <Text style={styles.successText}>
        Congratulations!
      </Text>
    </View>
  );

  return (
    <View style={styles.container}>
      <View style={styles.buttonCountainer} >
        <TouchableOpacity
          style={styles.clickerButton}
          onPress={() => incrementClicker()}
        >
          <Text testID="value" style={styles.clickerValue}>{clickerValue}</Text>
        </TouchableOpacity>
        {clickerValue < 10 && <Text style={styles.legend}>Please click on button</Text>}
      </View>

      <View style={{flex: 1}} testID="auto-clicker">
        {
          [...Array(config.clicker.autoClickerLevels)].map((e, i) => (
            <AutoClicker
              key={`Lvl ${i}`}
              name={`Lvl ${i}`}
              activeAutoClickers={autoClickers.filter((e) => e.name === `Lvl ${i}`).length}
              buyAmount={Math.floor(config.clicker.buyMinAmount * Math.exp(i))} // Crecimiento exponencial del costo base
              value={clickerValue}
              clickAmount={i + 1}
              onBuy={(name, clickAmount, buyAmount) => {
                createNewClicker(name, buyAmount, clickAmount);
              }}
              />
          ))
        }
      </View>
    </View>
  );
}

export default Clicker;
