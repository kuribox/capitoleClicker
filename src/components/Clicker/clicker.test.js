import React from 'react';
import renderer from 'react-test-renderer';
import Clicker from './index';
import TestContainer from '../../../test/TestContainer';
import {render, waitFor} from '@testing-library/react-native'

it('renders correctly', () => {
  const testRenderer = renderer.create(<TestContainer Component={Clicker} />);
  expect(testRenderer.toJSON()).toMatchSnapshot();
});

it('initial value', async () => {
  const { queryByTestId } = await waitFor(() => render(<TestContainer Component={Clicker} />));   
  const clickerValue = queryByTestId('value');
  expect(clickerValue.props.children).toBe(0);
});

it('storage value', async () => {
  const { queryByTestId } = await waitFor(() => render(<TestContainer Component={(props) => <Clicker user={{
    name: "name",
    value: 10,
    clickers: [],
  }} {...props}/>} />));

  const clickerValue = queryByTestId('value');
  expect(clickerValue.props.children).toBe(10);
});

it('autoClicker lvl 1', async () => {
  const { queryByTestId } = await waitFor(() => render(<TestContainer Component={(props) => <Clicker user={{
    name: "name",
    value: 10,
    clickers: [],
  }} {...props}/>} />));

  const autoClicker = queryByTestId('auto-clicker');
  
  await waitFor(() => {
    autoClicker.props.children[0].props.onBuy("lvl 1", 10, 1);
  });

  const clickerValue = queryByTestId('value');
  await waitFor(() => {
    expect(clickerValue.props.children).toBeGreaterThan(10);
  }); 
});


