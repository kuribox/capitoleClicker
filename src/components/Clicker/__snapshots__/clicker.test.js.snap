// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "space-between",
      "padding": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#072146",
          "borderColor": "#32507c",
          "borderRadius": 60,
          "borderWidth": 10,
          "marginVertical": 10,
          "opacity": 1,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 35,
            "padding": 10,
            "textAlign": "center",
          }
        }
        testID="value"
      >
        0
      </Text>
    </View>
    <Text
      style={
        Object {
          "fontSize": 18,
        }
      }
    >
      Please click on button
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
    testID="auto-clicker"
  />
</View>
`;
